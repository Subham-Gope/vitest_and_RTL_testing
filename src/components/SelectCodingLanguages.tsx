import React, { useState } from "react";
import type { ChangeEvent } from "react";

const SelectCodingLanguages: React.FC = () => {
  const [value, setValue] = useState<string[]>([]);

  const handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
    const selectedOptions = Array.from(
      e.target.selectedOptions,
      (option) => option.value
    );

    setValue(selectedOptions);
  };

  return (
    <>
      <div>
        <label htmlFor="coding-lang">Select your coding languages:</label>
        <select id="coding-lang" value={value} onChange={handleChange} multiple>
          <option value="" disabled>
            Selected
          </option>
          <option value="html">HTML</option>
          <option value="css">CSS</option>
          <option value="javascript">javaScript</option>
          <option value="java">Java</option>
          <option value="rust">Rust</option>
          <option value="typescript">TypeScript</option>
        </select>
      </div>
      <div>
        <span className="text-3xl text-gray-800">Selected Languages:</span>
        <ul>
          {value.map((lang, index) => (
            <li key={index}>
              <p>{lang}</p>
            </li>
          ))}
        </ul>
      </div>
    </>
  );
};

export default SelectCodingLanguages;
