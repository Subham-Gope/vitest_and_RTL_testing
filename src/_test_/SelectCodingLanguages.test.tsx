import { describe, test, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SelectCodingLanguages from "../components/SelectCodingLanguages";

describe("Testing the SelectCodingLanguages.tsx file", () => {
  test("selects multiple options and deselects some from the select box", async () => {
    // setting up the user setup()
    const user = userEvent.setup();

    // setting up the debugger
    screen.debug();

    // setting up the snapshot matching
    const { container } = render(<SelectCodingLanguages />);
    expect(container).toMatchSnapshot();

    const select = screen.getByRole("listbox", {
      name: /select your coding languages:/i,
    });
    expect(select).toBeInTheDocument();

    // Initiating user interactions
    await user.selectOptions(select, ["HTML", "CSS"]);

    const htmlOption = screen.getByRole("option", {
      name: /html/i,
    }) as HTMLOptionElement;
    const cssOption = screen.getByRole("option", {
      name: /css/i,
    }) as HTMLOptionElement;
    expect(htmlOption).toBeInTheDocument();
    expect(cssOption).toBeInTheDocument();
    expect(htmlOption.selected).toBe(true);
    expect(cssOption.selected).toBe(true);
  });
});
