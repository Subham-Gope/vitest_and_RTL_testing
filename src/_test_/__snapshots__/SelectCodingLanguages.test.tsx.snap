// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the SelectCodingLanguages.tsx file > selects multiple options and deselects some from the select box 1`] = `
<div>
  <div>
    <label
      for="coding-lang"
    >
      Select your coding languages:
    </label>
    <select
      id="coding-lang"
      multiple=""
    >
      <option
        disabled=""
        value=""
      >
        Selected
      </option>
      <option
        value="html"
      >
        HTML
      </option>
      <option
        value="css"
      >
        CSS
      </option>
      <option
        value="javascript"
      >
        javaScript
      </option>
      <option
        value="java"
      >
        Java
      </option>
      <option
        value="rust"
      >
        Rust
      </option>
      <option
        value="typescript"
      >
        TypeScript
      </option>
    </select>
  </div>
  <div>
    <span
      class="text-3xl text-gray-800"
    >
      Selected Languages:
    </span>
    <ul />
  </div>
</div>
`;
